### PR TITLE
4dsound nap pr january 2020

### PR DESCRIPTION
**Pull request by stijnvanbeek**
These are a couple of changes to the NAP repository made by 4DSOUND:

these will concern Naivi most:
- The AppRunner shuts down the services when the initialization fails, to ensure shutdown() is always called when init() has succeeded.
- The Parameter::setValue() functions are exposed to rtti, so they can be called form python scripts.

And these less so, as they are mainly audio changes:
- The signature of audio::MultiInputPin::pull() method has changed, so the caller has the responsibility to manage the destination buffer. This solves some threadsafety concerns while parallel processing audio.
- The host API index is passed to AudioService::openStream() method, to make sure it is updated internally when the device settings change at runtime.
- A simplification in the SafeOwner/SafePtr/DeletionQueue system, that is possible now because in NAP 0.4 order of shutdown/destruction of services is now guaranteed to mirror the service dependency graph.
- An couple of adaptations in utility/threading.h: classes are marked with NAPAPI macro to use them across DLL bounds on windows, threadpools can be marked realtime, in order to spawn their threads with realtime priority.

### CODE REVIEW

**Comment by cklosters** [comment]
Thanks for the PR. Because auto PR building is not installed for the public repo it's hard to verify the changes cross platform. For now I enabled auto PR building for the public repo, because the number of PRs are limited. Best to push a change and see if the system picks it up. I can't let anything pass without cross-platform verification.

File: modules/napapp/src/apprunner.h [hunk 1]
```diff
@@ -174,7 +174,11 @@ namespace nap
 
 		// Initialize application
 		if(!error.check(app.init(error), "unable to initialize application"))
+		{
+			mCore.shutdownServices();
+			error.fail("Failed to initialize application");
```
Comment (cklosters) [inline modules/napapp/src/apprunner.h:179]: This message is redundant, please remove. Message is already added when checked

File: modules/napaudio/src/audio/core/audiopin.h [hunk 2]
```diff
@@ -153,9 +153,9 @@ namespace nap
 			
 			/**
 			 * This method can be used by the node to pull a buffer of samples for every connected output pin.
-			 * @return: the vector can contain nullptr items if somewhere down the line of connection silence is returned.
+			 * @param result the vector will be resized and filled with a SampleBuffer* for each connected pin. The vector can contain nullptr items if somewhere down the line of connection silence is returned. It is advised to allocate the result vector in the constructor of the Node class that contains the MultiInpuPin, and to preallocate memory using vector<>::reserve(), in order to avoid allocations on the audio thread.
 			 */
-			std::vector<SampleBuffer*>& pull();
+			void pull(std::vector<SampleBuffer*>& result);
```
Comment (cklosters) [inline modules/napaudio/src/audio/core/audiopin.h:158]: indenting issue

File: modules/napparameter/src/parameternumeric.h [hunk 3]
```diff
@@ -109,5 +109,7 @@ namespace nap
 			RTTI_PROPERTY("Value",		&Type::mValue,			nap::rtti::EPropertyMetaData::Default)				\
 			RTTI_PROPERTY("Minimum",	&Type::mMinimum,		nap::rtti::EPropertyMetaData::Default)				\
 			RTTI_PROPERTY("Maximum",	&Type::mMaximum,		nap::rtti::EPropertyMetaData::Default)				\
+			RTTI_FUNCTION("setValue",	static_cast<void (Type::*)(decltype(Type::mValue))>(&Type::setValue))	\
```
Comment (cklosters) [inline modules/napparameter/src/parameternumeric.h:112]: We should really start thinking about moving this to specialized interfaces for Python, we can't keep scattering around function definitions because you need them for this particular case. We either support it for a certain class or don't.

File: utility/src/utility/threading.cpp [hunk 4]
```diff
@@ -1,13 +1,13 @@
...
 
+#ifdef _WIN32
+#include <windows.h>
+#endif
+
 namespace nap
 {
     
-    TaskQueue::TaskQueue(std::uint32_t maxQueueItems) : mQueue(maxQueueItems)
+    TaskQueue::TaskQueue(unsigned int maxQueueItems) : mQueue(maxQueueItems)
```
Comment (cklosters) [inline utility/src/utility/threading.cpp:10]: why be less explicit here? also: use nap/numeric

File: utility/src/utility/threading.cpp [hunk 5]
```diff
@@ -16,23 +16,27 @@ namespace nap
...
+            for (auto i = 0; i < count; ++i)
+                (*it++)();
+            it = mDequeuedTasks.begin();
+            count = mQueue.try_dequeue_bulk(it, mDequeuedTasks.size());
+        }
     }
     
     
-    WorkerThread::WorkerThread(bool blocking, std::uint32_t maxQueueItems) : mBlocking(blocking), mTaskQueue(maxQueueItems)
+    WorkerThread::WorkerThread(bool blocking, unsigned int maxQueueItems) : mBlocking(blocking), mTaskQueue(maxQueueItems)
```
Comment (cklosters) [inline utility/src/utility/threading.cpp:39]: Same, why less explicit?

File: utility/src/utility/threading.h [hunk 6]
```diff
@@ -1,12 +1,8 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (cklosters) [inline utility/src/utility/threading.h:1]: Why is this removed?

File: utility/src/utility/threading.h [hunk 7]
```diff
@@ -102,16 +99,19 @@ namespace nap
     /**
      * A pool of threads that can be used to perform multiple tasks at the same time
      */
-    class ThreadPool final
+    class NAPAPI ThreadPool final
 	{
     public:
-        ThreadPool(std::uint32_t numberOfThreads = 1, std::uint32_t maxQueueItems = 20);
+        ThreadPool(unsigned int numberOfThreads = 1, unsigned int maxQueueItems = 20, bool realTimePriority = false);
```
Comment (cklosters) [inline utility/src/utility/threading.h:105]: Same, why less explicit?

File: modules/napparameter/src/parameternumeric.h (hunk 3 above)
Comment (stijnvanbeek) [reply modules/napparameter/src/parameternumeric.h:112]: I agree. There is the design document that I wrote about this:
https://docs.google.com/document/d/1QbV9nd_7S-vZH2Vq0V5KqjmP1FCYyfy5-ZzGA4Vd3Kw
For now: let's please support it for Parameter classes.

File: utility/src/utility/threading.cpp (hunk 4 above)
Comment (stijnvanbeek) [reply utility/src/utility/threading.cpp:14]: I removed all the unsigned definitions and just used "int" here. I think it should be just an ordinary int argument.

**Comment by cklosters** [comment]
```
[18:38:15]	[Step 2/2] naputility.lib(threading.obj) : warning LNK4217: locally defined symbol ?enqueue@TaskQueue@nap@@QEAAXV?$function@$$A6AXXZ@std@@@Z (public: void __cdecl nap::TaskQueue::enqueue(class std::function<void __cdecl(void)>)) imported in function "public: virtual __cdecl nap::WorkerThread::~WorkerThread(void)" (??1WorkerThread@nap@@UEAA@XZ) [C:\Agent\work\44040c56bd2f12b7\nap\packaging_build\modules\napaudio\mod_napaudio.vcxproj]
[18:38:15]	[Step 2/2] naputility.lib(threading.obj) : warning LNK4217: locally defined symbol ??1TaskQueue@nap@@QEAA@XZ (public: __cdecl nap::TaskQueue::~TaskQueue(void)) imported in function "int `public: __cdecl nap::ThreadPool::ThreadPool(int,int,bool)'::`1'::dtor$4" (?dtor$4@?0???0ThreadPool@nap@@QEAA@HH_N@Z@4HA) [C:\Agent\work\44040c56bd2f12b7\nap\packaging_build\modules\napaudio\mod_napaudio.vcxproj]
[18:38:15]	[Step 2/2] naputility.lib(threading.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) const nap::WorkerThread::`vftable'" (__imp_??_7WorkerThread@nap@@6B@) referenced in function "public: __cdecl nap::WorkerThread::WorkerThread(bool,int)" (??0WorkerThread@nap@@QEAA@_NH@Z) [C:\Agent\work\44040c56bd2f12b7\nap\packaging_build\modules\napaudio\mod_napaudio.vcxproj]
[18:38:15]	[Step 2/2] C:\Agent\work\44040c56bd2f12b7\nap\packaging_bin\MSVC-x86_64-Release\mod_napaudio.dll : fatal error LNK1120: 1 unresolved externals [C:\Agent\work\44040c56bd2f12b7\nap\packaging_build\modules\napaudio\mod_napaudio.vcxproj]
```

**Comment by cklosters** [comment]
```
[18:32:32]	[Step 2/2] C:\Agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.cpp(14): warning C4273: 'nap::TaskQueue::TaskQueue': inconsistent dll linkage [C:\Agent\work\44040c56bd2f12b7\nap\packaging_build\utility\naputility.vcxproj]
[18:32:32]	[Step 2/2] c:\agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.h(32): note: see previous definition of '{ctor}' (compiling source file C:\Agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.cpp)
[18:32:32]	[Step 2/2] C:\Agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.cpp(22): warning C4273: 'nap::TaskQueue::processBlocking': inconsistent dll linkage [C:\Agent\work\44040c56bd2f12b7\nap\packaging_build\utility\naputility.vcxproj]
[18:32:32]	[Step 2/2] c:\agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.h(42): note: see previous definition of 'processBlocking' (compiling source file C:\Agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.cpp)
[18:32:32]	[Step 2/2] C:\Agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.cpp(30): warning C4273: 'nap::TaskQueue::process': inconsistent dll linkage [C:\Agent\work\44040c56bd2f12b7\nap\packaging_build\utility\naputility.vcxproj]
[18:32:32]	[Step 2/2] c:\agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.h(47): note: see previous definition of 'process' (compiling source file C:\Agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.cpp)
[18:32:32]	[Step 2/2] C:\Agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.cpp(43): warning C4273: 'nap::WorkerThread::WorkerThread': inconsistent dll linkage [C:\Agent\work\44040c56bd2f12b7\nap\packaging_build\utility\naputility.vcxproj]
[18:32:32]	[Step 2/2] c:\agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.h(67): note: see previous definition of '{ctor}' (compiling source file C:\Agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.cpp)
[18:32:32]	[Step 2/2] C:\Agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.cpp(50): warning C4273: 'nap::WorkerThread::~WorkerThread': inconsistent dll linkage [C:\Agent\work\44040c56bd2f12b7\nap\packaging_build\utility\naputility.vcxproj]
[18:32:32]	[Step 2/2] c:\agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.h(68): note: see previous definition of '{dtor}' (compiling source file C:\Agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.cpp)
[18:32:32]	[Step 2/2] C:\Agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.cpp(56): warning C4273: 'nap::WorkerThread::start': inconsistent dll linkage [C:\Agent\work\44040c56bd2f12b7\nap\packaging_build\utility\naputility.vcxproj]
[18:32:32]	[Step 2/2] c:\agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.h(78): note: see previous definition of 'start' (compiling source file C:\Agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.cpp)
[18:32:32]	[Step 2/2] C:\Agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.cpp(81): warning C4273: 'nap::WorkerThread::stop': inconsistent dll linkage [C:\Agent\work\44040c56bd2f12b7\nap\packaging_build\utility\naputility.vcxproj]
[18:32:32]	[Step 2/2] c:\agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.h(83): note: see previous definition of 'stop' (compiling source file C:\Agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.cpp)
[18:32:32]	[Step 2/2] C:\Agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.cpp(95): warning C4273: 'nap::ThreadPool::ThreadPool': inconsistent dll linkage [C:\Agent\work\44040c56bd2f12b7\nap\packaging_build\utility\naputility.vcxproj]
[18:32:32]	[Step 2/2] c:\agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.h(109): note: see previous definition of '{ctor}' (compiling source file C:\Agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.cpp)
[18:32:32]	[Step 2/2] C:\Agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.cpp(104): warning C4273: 'nap::ThreadPool::~ThreadPool': inconsistent dll linkage [C:\Agent\work\44040c56bd2f12b7\nap\packaging_build\utility\naputility.vcxproj]
[18:32:32]	[Step 2/2] c:\agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.h(110): note: see previous definition of '{dtor}' (compiling source file C:\Agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.cpp)
[18:32:32]	[Step 2/2] C:\Agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.cpp(110): warning C4273: 'nap::ThreadPool::shutDown': inconsistent dll linkage [C:\Agent\work\44040c56bd2f12b7\nap\packaging_build\utility\naputility.vcxproj]
[18:32:32]	[Step 2/2] c:\agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.h(123): note: see previous definition of 'shutDown' (compiling source file C:\Agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.cpp)
[18:32:32]	[Step 2/2] C:\Agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.cpp(125): warning C4273: 'nap::ThreadPool::resize': inconsistent dll linkage [C:\Agent\work\44040c56bd2f12b7\nap\packaging_build\utility\naputility.vcxproj]
[18:32:32]	[Step 2/2] c:\agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.h(128): note: see previous definition of 'resize' (compiling source file C:\Agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.cpp)
[18:32:32]	[Step 2/2] C:\Agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.cpp(135): warning C4273: 'nap::ThreadPool::addThread': inconsistent dll linkage [C:\Agent\work\44040c56bd2f12b7\nap\packaging_build\utility\naputility.vcxproj]
[18:32:32]	[Step 2/2] c:\agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.h(141): note: see previous definition of 'addThread' (compiling source file C:\Agent\work\44040c56bd2f12b7\nap\utility\src\utility\threading.cpp)
```

**Comment by cklosters** [comment]
I'm still seeing the new warnings on windows:

`'nap::TaskQueue::TaskQueue': inconsistent dll linkage`

Also, what does removing the NAPAPI macro actually solve in theory? We need external linkage, therefore we export the symbols. Can you test these changes on a windows machine on your end and fix them?

**Comment by cklosters** [comment]
Moving base of PR to public_beta. All PRs will be merged into public_beta until ready for release (tested).

File: utility/src/utility/threading.cpp (hunk 4 above)
Comment (cklosters) [reply utility/src/utility/threading.cpp]: use int when accuracy is of no importance at all, ie: you know it will be in bounds, always. Use nap::uint32 etc. when you want to be explicit and show intention.
